### ISSUE TITLE
Spikekill fills in NaN values outside specified time range

### ISSUE THREAD

**Issue by riversdev0** [Cacti/cacti#3675]
# Describe the bug

When running Spikekill using GapFill method on a graph containing holes in the data, Spikekill will fill in values that are outside the time range.

## To Reproduce

1. Find a graph with holes in the data.  Here is my example graph.  (Red arch line added.)
![image](https://user-images.githubusercontent.com/7363202/87082407-9b4c4480-c1f0-11ea-9c6c-5147064758ca.png)

2. Zoom into the graph at a time range that contains a hole, but not the first several holes.  In the example above, I zoomed into the range where I drew the red arch.

3. Run Spikekill using the GapFill method.

4. Observe fixed graph.  This looks good so far because the hole was filled in and the spike was mitigated.
![image](https://user-images.githubusercontent.com/7363202/87082694-072ead00-c1f1-11ea-95ff-7acb17b1abb7.png)

5. Zoom out of the 'fixed' graph and observe what happened.  I drew the red arch once again for you to see that my hole is filled in and the spike is mitigated.  However, there are many holes in the left part of the graph that are suddenly filled in, despite being outside the time range that was specified.
![image](https://user-images.githubusercontent.com/7363202/87083130-b79cb100-c1f1-11ea-80fb-084180ea56e2.png)

6. Since I had Cacti's debug level set to Debug, I can observe in the log file that the time range is limited to the range of the graph above (which is perfect!).  I can also observe in the log that Spikekill mitigated 14 spikes, which is far more than existed in my time range.
```
2020/07/09 14:24:21 - SPIKEKILL CactiUser:admin, File:ping_int_velocity_tunb_ping_1282.rrd, Method:GapFill, OutStart:1594060746, OutEnd:1594061254, AvgNan:last
2020/07/09 14:24:22 - SPIKE DEBUG: number of RRAs: 12
2020/07/09 14:24:22 - SPIKE DEBUG: number of DSes: 1
2020/07/09 14:24:22 - SPIKE DEBUG: Limited to window: Jul 6, 2020 13:39:06 thru Jul 6, 2020 13:47:34
2020/07/09 14:24:22 - SPIKEKILL DEBUG: replacing dsvalue NaN with last value 2.7600000000e+01
2020/07/09 14:24:22 - SPIKEKILL DEBUG: replacing dsvalue NaN with last value 2.8545000000e+01
2020/07/09 14:24:22 - SPIKEKILL DEBUG: replacing dsvalue NaN with last value 2.8830000000e+01
2020/07/09 14:24:22 - SPIKEKILL DEBUG: replacing dsvalue NaN with last value 2.7500000000e+01
2020/07/09 14:24:22 - SPIKEKILL DEBUG: replacing dsvalue NaN with last value 2.7570000000e+01
2020/07/09 14:24:22 - SPIKEKILL DEBUG: replacing dsvalue 5.9415000000e+01 with last value 2.7636666667e+01
2020/07/09 14:24:22 - SPIKEKILL DEBUG: replacing dsvalue NaN with last value 2.7636666667e+01
2020/07/09 14:24:22 - SPIKEKILL DEBUG: replacing dsvalue 5.9415000000e+01 with last value 2.7636666667e+01
2020/07/09 14:24:22 - SPIKEKILL DEBUG: replacing dsvalue NaN with last value 2.7636666667e+01
2020/07/09 14:24:22 - SPIKEKILL DEBUG: replacing dsvalue 5.9415000000e+01 with last value 2.7636666667e+01
2020/07/09 14:24:22 - SPIKEKILL DEBUG: replacing dsvalue NaN with last value 2.7636666667e+01
2020/07/09 14:24:23 - SPIKEKILL DEBUG: replacing dsvalue 5.9415000000e+01 with last value 2.7800000000e+01
2020/07/09 14:24:23 - SPIKEKILL DEBUG: replacing dsvalue 5.9415000000e+01 with last value 2.7636666667e+01
2020/07/09 14:24:23 - SPIKEKILL DEBUG: replacing dsvalue NaN with last value 2.7636666667e+01
2020/07/09 14:24:24 - WEBUI WARNING: Removed '14' Spikes from '/var/www/html/cacti/rra/ping_int_velocity_tunb_ping_1282.rrd', Method:'4'
```

## Expected behavior

I would expect that the time range provided to Spikekill would prevent any modifications of data outside the given time range.

## Screenshots

Provided above.

## Desktop (please complete the following information)

n/a

## Smartphone (please complete the following information)

n/a

## Additional context

None

**Comment by TheWitness** [comment]
Change the logging to include the time values for that interval from the database to get more data.  @netniV may have to merge the open pull request for you to get the latest changes though.

**Comment by TheWitness** [comment]
That's also looking like 'float' and not gapfill.  Can you confirm the callback too?

**Comment by TheWitness** [comment]
And also using 'last' value, which should be fixed now.

**Comment by riversdev0** [comment]
> That's also looking like 'float' and not gapfill. Can you confirm the callback too?

The log output agrees with the option I clicked in the GUI.  According to the code, running GapFill not only fills in missing data points, but also kills spikes, using the Variance Average method.  In contrast, when you run Float, it overwrites valid data points without regard for whether their values fall inline with average or not.  So, with regard to whether this is GapFill or Float, that much seems to be working fine.

As an aside, when I wrote the SpikeKill documentation, I included in there that GapFill also searches for spikes.  If that is not the intended behavior (or if it is not desirable behavior), we can change the SpikeKill script and the associated documentation.  However, I haven't submitted a PR for my re-write of SpikeKill yet because I wanted to finish the LDAP documentation.

> Change the logging to include the time values for that interval from the database to get more data.

That is a good idea, however, I know this bug is verified because I actually discovered it by reading the raw code rather than actually attempting the operation above.  When I perceived in the code that this could be a problem, I located a graph which would elicit the issue, and was able to verify the behavior to grab the screenshots above.

I am not far from having a fixed version of the SpikeKill script ready, but there are going to be several major changes.  Therefore, I wanted to have these issues documented to make it easier for you guys to review the changes in the PR.  I hope this was the right way for me to be collaborating.

**Comment by TheWitness** [comment]
@riversdev0, I've added a 7/30/2020 date for the next point release (1.2.14).  Continue working with those timeframes

**Comment by riversdev0** [comment]
I think @netniV wants this against 'develop', right, so does that make the deadline moot?  I think i can get this done by 7/30, but i'm going to be out of pocket for the next 6 days.  I should have something by the end of next week.

**Comment by TheWitness** [comment]
@riversdev0, Do what you can.  Life is short.  Thanks for your help.

**Comment by riversdev0** [comment]
@netniV + @TheWitness , I'm nearing completion of the updated script.  Would you both like it under 'develop' or '1.2.x'?  The code changes are somewhat extensive, but the actual functionality is the same.  This cleaned up version employs a more predictable logic flow to ensure that unexpected things (like this issue #3675) don't happen anymore.  There is one new enhancement introduced, which relates to issue #3620 .  I'll continue my testing, meanwhile please let me know what branch to use for the PR.

**Comment by netniV** [comment]
I think it should have some testing time if the changes are extensive and that should against 1.3 not 1.2.x.  Small patches for bug fixes should be against 1.2.x, large changes like this should be against develop.
